proposals/XXXX-key-paths.md

```diff
@@ -0,0 +1,175 @@
+# Smart KeyPaths: Better Key-Value Coding for Swift
+
+* Proposal: SE-NNNN
+* Authors: [David Smith](https://github.com/Catfish-Man), [Michael LeHew](https://github.com/mlehew), [Joe Groff](https://github.com/jckarter)
+* Review Manager: TBD
+* Status: **Awaiting Review**
+* Associated PRs:
+   * [#644](https://github.com/apple/swift-evolution/pull/644)
+
+## Introduction
+We propose a family of concrete _Key Path_ types that represent uninvoked references to properties that can be composed to form paths through many values and directly get/set their underlying values.
+
+## Motivation
+#### We Can Do Better than String
+On Darwin platforms Swift's existing `#keyPath()` syntax provides a convenient way to safely *refer* to properties. Unfortunately, once validated, the expression becomes a `String` which has a number of important limitations:
+
+* Loss of type information (requiring awkward `Any` APIs)
+* Unnecessarily slow to parse
+* Only applicable to `NSObjects` 
+* Limited to Darwin platforms
+
+#### Use/Mention Distinctions
+While methods can be referred to without invoking them (`let x = foo.bar` instead of  `let x = foo.bar()`), this is not currently possible for properties and subscripts.
+
+Making indirect references to a properties' concrete types also lets us expose metadata about the property, and in the future additional behaviors.
+
+#### More Expressive KeyPaths
+We would also like to support being able to use _Key Paths_ to access into collections, which is not currently possible.
+
+## Proposed solution
+We propose introducing a new expression akin to `Type.method`, but for properties and subscripts. These property reference expressions produce `KeyPath` objects, rather than `Strings`. `KeyPaths` are a family of generic classes _(structs and protocols here would be ideal, but requires generalized existentials)_ which encapsulate a property reference or chain of property references, including the type, mutability, property name(s), and ability to set/get values.
+
+Here's a sample of it in use:
+
+```swift
+struct Person {
+	var name: String
+	var friends: [Person]
+	var bestFriend: Person?
+}
+
+var han = Person(name: "Han Solo", friends: [])
+var luke = Person(name: "Luke Skywalker", friends: [han])
+
+let firstFriendsNameKeyPath = Person.friends[0].name
+
+let firstFriend = luke[path] // han
```

Comment:
Yes it should.  That was a Last Minute Change™ where the find-replace failed. :(